components/raftstore-v2/src/operation/query/local.rs

```diff
@@ -2,44 +2,251 @@
+                RequestPolicy::StaleRead => {
+                    let read_ts = decode_u64(&mut req.get_header().get_flag_data()).unwrap();
+                    delegate.check_stale_read_safe(read_ts)?;
+
+                    let region = Arc::clone(&delegate.region);
+                    let snap = RegionSnapshot::from_snapshot(
+                        delegate.get_snapshot(None, &mut None),
+                        region,
+                    );
+
+                    TLS_LOCAL_READ_METRICS.with(|m| m.borrow_mut().local_executed_requests.inc());
 
-use crate::{fsm::StoreMeta, tablet::CachedTablet};
+                    delegate.check_stale_read_safe(read_ts)?;
+
+                    TLS_LOCAL_READ_METRICS
+                        .with(|m| m.borrow_mut().local_executed_stale_read_requests.inc());
+                    Ok(Some(snap))
+                }
+                _ => unreachable!(),
+            },
+            Ok(None) => Ok(None),
+            Err(e) => {
+                let mut response = cmd_resp::new_error(e);
+                if let Some(delegate) = self
+                    .local_reader
+                    .delegates
+                    .get(&req.get_header().get_region_id())
+                {
+                    cmd_resp::bind_term(&mut response, delegate.term);
+                }
+                Err(response)
+            }
+        }
+    }
+
+    pub async fn snapshot(
+        &mut self,
+        mut req: RaftCmdRequest,
+    ) -> std::result::Result<RegionSnapshot<E::Snapshot>, RaftCmdResponse> {
+        let region_id = req.header.get_ref().region_id;
+        if let Some(snap) = self.try_get_snapshot(req.clone())? {
+            return Ok(snap);
+        }
+
+        if let Some(query_res) = self.try_to_renew_lease(region_id, &req).await? {
+            // If query successful, try again.
+            if query_res.read().is_some() {
+                req.mut_header().set_read_quorum(false);
+                if let Some(snap) = self.try_get_snapshot(req)? {
+                    return Ok(snap);
+                }
+            }
+        }
+
+        let mut err = errorpb::Error::default();
+        err.set_message(format!(
+            "Fail to get snapshot from LocalReader for region {}. Maybe due to `not leader` or `not applied to the current term`",
+            region_id
+        ));
+        let mut resp = RaftCmdResponse::default();
+        resp.mut_header().set_error(err);
+        Err(resp)
+    }
+
+    // try to renew the lease by sending read query where the reading process may
+    // renew the lease
+    async fn try_to_renew_lease(
+        &self,
+        region_id: u64,
+        req: &RaftCmdRequest,
+    ) -> std::result::Result<Option<QueryResult>, RaftCmdResponse> {
+        let (msg, sub) = PeerMsg::raft_query(req.clone());
+        let mut err = errorpb::Error::default();
+        match MsgRouter::send(&self.router, region_id, msg) {
+            Ok(()) => return Ok(sub.result().await),
```

Comment:
I'm wondering if we can clone and cache sub here. 
So when sub exists,  we don't need to issue the MsgRouter::send again for later requests that are before sub.result().await return . Just wait on sub.result().await. And once await returns, clear the cache. 
This is to dedup unnecessary renew lease requests to raftstore. 
With this, it can even remove the read index amend logic in raftstore v2. Because redundant read index messages will be greatly reduced or in fact completely removed. @BusyJay What do you think?

Reply:
If it is possible, I think we can propose another PR to implement it and see the improvements.

Reply:
ReadIndex amend works for requests across threads.